### ISSUE TITLE
undefined symbol: sormqr_

### ISSUE THREAD

**Issue by tudor-berariu** [pytorch/pytorch#507]
I compiled `pytorch` (using both Python 2 and Python 3) with CUDA support on Ubuntu 16.04 having gcc 4.8.5. The install process terminated with no critical errors.

Torch is already installed on the same machine and works fine.

When importing `pytorch` I get the following error:

    ----> 1 import torch
    
    /home/tudor/anaconda2/lib/python2.7/site-packages/torch/__init__.pyc in <module>()
         43 sys.setdlopenflags(_dl_flags.RTLD_GLOBAL | _dl_flags.RTLD_NOW)
         44 
    ---> 45 from torch._C import *
         46 
         47 __all__ += [name for name in dir(_C)
    
    ImportError: /home/tudor/anaconda2/lib/python2.7/site-packages/torch/lib/libTHC.so.1: undefined symbol: sormqr_

**Comment by soumith** [comment]
it looks like you used an anaconda environment.
Did you run the commands in order from here: https://github.com/pytorch/pytorch#from-source ? including the CMAKE_PREFIX_PATH command and installing mkl ahead of time...

**Comment by tudor-berariu** [comment]
Yes, I followed all steps.

**Comment by apaszke** [comment]
It looks like a MAGMA issue, can you try and check what THC's CMake logged when configuring the build (only the lines about MAGMA)?

**Comment by soumith** [comment]
if you are using CUDA 7.5, be sure to do:
```
conda install -c soumith magma-cuda75
```

and if you are using CUDA 8.0, do:
```
conda install -c soumith magma-cuda80
```

**Comment by tudor-berariu** [comment]
@soumith Yes i did install the correct version for `magma-cuda`.

@apaszke Not sure if this is the relevant info, but the only lines in `pytorch/torch/lib/build/THC/CMakeFiles/CMakeOutput.log` referring to magma are:

    Feature record: CXX_FEATURE:0cxx_variadic_macros
    Feature record: CXX_FEATURE:0cxx_variadic_templates
    Determining if the prototype magma_get_sgeqrf_nb exists for MAGMA_V2 passed with the following output:
    Change Dir: /home/tudor/pytorch/torch/lib/build/THC/CMakeFiles/CMakeTmp
    
    Run Build Command:"/usr/bin/make" "cmTC_2e3cf/fast"

**Comment by soumith** [comment]
@tudor-berariu thanks. what's your CMake version?

**Comment by apaszke** [comment]
@tudor-berariu Could you grep for some of [these prints](https://github.com/pytorch/pytorch/blob/master/torch/lib/THC/CMakeLists.txt#L65-L68)? It might be necessary to clean and rebuild THC 😕

**Comment by tudor-berariu** [comment]
@soumith cmake version 3.6.3

@apaszke 

```
$ python setup.py install | grep -i magma
- Compiling with MAGMA support
-- MAGMA INCLUDE DIRECTORIES: /home/tudor/anaconda2/include
-- MAGMA LIBRARIES: /home/tudor/anaconda2/lib/libmagma.a
-- MAGMA V2 check: 1
[ 15%] Building NVCC (Device) object CMakeFiles/THC.dir/THC_generated_THCTensorMathMagma.cu.o
-- Up-to-date: /home/tudor/pytorch/torch/lib/tmp_install/include/THC/THCTensorMathMagma.cuh
-- Up-to-date: /home/tudor/pytorch/torch/lib/tmp_install/include/THC/generic/THCTensorMathMagma.h
-- Up-to-date: /home/tudor/pytorch/torch/lib/tmp_install/include/THC/generic/THCTensorMathMagma.cu
copying torch/lib/include/THC/generic/THCTensorMathMagma.h -> build/lib.linux-x86_64-2.7/torch/lib/include/THC/generic
copying build/lib.linux-x86_64-2.7/torch/lib/include/THC/generic/THCTensorMathMagma.h -> /home/tudor/anaconda2/lib/python2.7/site-packages/torch/lib/include/THC/generic
```

**Comment by apaszke** [comment]
weird, it seems to be picked up correctly. can you prepend `VERBOSE=1` to the install command, so we can see the compiler commands used to build the lib? Maybe it doesn't get linked for some reason...

**Comment by tudor-berariu** [comment]
[This](https://gist.github.com/tudor-berariu/4527bad7a1c649bfedce0e92990d1953) is the full output of the install process (with `VERBOSE=1`). I still can't figure out what's wrong.

**Comment by apaszke** [comment]
Can you try `nm /home/tudor/anaconda2/lib/libmagma.a | grep ormqr`?

**Comment by apaszke** [comment]
Hm false alert. It looks like a TH issue. The only place where `sormqr_` is used is [there](https://github.com/pytorch/pytorch/blob/0f65c9267d5ec55584b0ec65acb5374c95af9c16/torch/lib/TH/generic/THLapack.c#L246).

**Comment by tudor-berariu** [comment]
@apaszke Thanks a lot for investigating this. The output of the grep command is [here](https://gist.github.com/tudor-berariu/dd89b1a98d1bbb5809c0669400136973).

**Comment by apaszke** [comment]
Ok! This is helpful. It seems that you don't have any LAPACK library, while your `libmagma.a` requires to be linked together with one. Did you install MKL (`conda install mkl`)?

**Comment by soumith** [comment]
I would suggest removing mkl and installing it again.
From the logs, it looks like you have a partial install of MKL somehow.

**Comment by tudor-berariu** [comment]
I did install `mkl`. Actually I've installed everything from scratch several times before posting the first time. I'll try to see what's wrong with `mkl`. What's interesting is that installing the pytorch binaries using conda works (of course without having CUDA support).

Thank you very much for your time. I appreciate your help.

**Comment by apaszke** [comment]
The problem is that TH can't find MKL when it's built. Where is it installed? Can you add the path (without `/lib` suffix) to `CMAKE_PREFIX_PATH`?

**Comment by soumith** [comment]
i'm leaning towards a CMake bug.
From tudor's logs this is the critical piece:

```bash
-- Checking for [mkl_gf_lp64 - mkl_gnu_thread - mkl_core - pthread - m]
--   Library mkl_gf_lp64: not found
-- Checking for [mkl_gf_lp64 - mkl_intel_thread - mkl_core - pthread - m]
--   Library mkl_gf_lp64: not found
-- Checking for [mkl_gf - mkl_gnu_thread - mkl_core - pthread - m]
--   Library mkl_gf: not found
-- Checking for [mkl_gf - mkl_intel_thread - mkl_core - pthread - m]
--   Library mkl_gf: not found
-- Checking for [mkl_intel_lp64 - mkl_gnu_thread - mkl_core - pthread - m]
--   Library mkl_intel_lp64: /home/tudor/anaconda2/lib/libmkl_intel_lp64.so
--   Library mkl_gnu_thread: not found
-- Checking for [mkl_intel_lp64 - mkl_intel_thread - mkl_core - pthread - m]
--   Library mkl_intel_lp64: /home/tudor/anaconda2/lib/libmkl_intel_lp64.so
--   Library mkl_intel_thread: /home/tudor/anaconda2/lib/libmkl_intel_thread.so
--   Library mkl_core: /home/tudor/anaconda2/lib/libmkl_core.so
--   Library pthread: not found
-- Checking for [mkl_intel - mkl_gnu_thread - mkl_core - pthread - m]
--   Library mkl_intel: not found
-- Checking for [mkl_intel - mkl_intel_thread - mkl_core - pthread - m]
--   Library mkl_intel: not found
-- Checking for [mkl_gf_lp64 - mkl_sequential - mkl_core - m]
--   Library mkl_gf_lp64: not found
-- Checking for [mkl_gf - mkl_sequential - mkl_core - m]
--   Library mkl_gf: not found
-- Checking for [mkl_intel_lp64 - mkl_sequential - mkl_core - m]
--   Library mkl_intel_lp64: /home/tudor/anaconda2/lib/libmkl_intel_lp64.so
--   Library mkl_sequential: /home/tudor/anaconda2/lib/libmkl_sequential.so
--   Library mkl_core: /home/tudor/anaconda2/lib/libmkl_core.so
--   Library m: not found
-- Checking for [mkl_intel - mkl_sequential - mkl_core - m]
--   Library mkl_intel: not found
-- Checking for [mkl_gf_lp64 - mkl_gnu_thread - mkl_core - pthread - m]
--   Library mkl_gf_lp64: not found
-- Checking for [mkl_gf_lp64 - mkl_intel_thread - mkl_core - pthread - m]
--   Library mkl_gf_lp64: not found
-- Checking for [mkl_gf - mkl_gnu_thread - mkl_core - pthread - m]
--   Library mkl_gf: not found
-- Checking for [mkl_gf - mkl_intel_thread - mkl_core - pthread - m]
--   Library mkl_gf: not found
-- Checking for [mkl_intel_lp64 - mkl_gnu_thread - mkl_core - pthread - m]
--   Library mkl_intel_lp64: /home/tudor/anaconda2/lib/libmkl_intel_lp64.so
--   Library mkl_gnu_thread: not found
-- Checking for [mkl_intel_lp64 - mkl_intel_thread - mkl_core - pthread - m]
--   Library mkl_intel_lp64: /home/tudor/anaconda2/lib/libmkl_intel_lp64.so
--   Library mkl_intel_thread: /home/tudor/anaconda2/lib/libmkl_intel_thread.so
--   Library mkl_core: /home/tudor/anaconda2/lib/libmkl_core.so
--   Library pthread: not found
-- Checking for [mkl_intel - mkl_gnu_thread - mkl_core - pthread - m]
--   Library mkl_intel: not found
-- Checking for [mkl_intel - mkl_intel_thread - mkl_core - pthread - m]
--   Library mkl_intel: not found
-- Checking for [mkl - guide - pthread - m]
--   Library mkl: not found
```

It finds some MKL libs but not all.

I'm trying to reproduce it with the same version of cmake that he has.

**Comment by apaszke** [comment]
It finds all of MKL, only pthread is missing:
```
-- Checking for [mkl_intel_lp64 - mkl_intel_thread - mkl_core - pthread - m]
--   Library mkl_intel_lp64: /home/tudor/anaconda2/lib/libmkl_intel_lp64.so
--   Library mkl_intel_thread: /home/tudor/anaconda2/lib/libmkl_intel_thread.so
--   Library mkl_core: /home/tudor/anaconda2/lib/libmkl_core.so
--   Library pthread: not found
```

**Comment by soumith** [comment]
also, libm is not found, it kinda needs to be there :
`
--   Library m: not found
`

wonder why both are missing.
Usually libm is located in `/usr/lib64/libm.so` and so will pthread be.

@tudor-berariu what machine are you running this on? what OS? what version? what Distro?

**Comment by tudor-berariu** [comment]
I'm running on an up-to-date Ubuntu 16.04.01 LTS, kernel 4.4.0-59 @ Lenovo Y50.
Cuda 8.0.44; Nvidia drivers 367.57 @ GeForce GTX 860M
gcc 4.8.5

You were right, the math and pthread libraries were the missing and they were not linked. On my machine the two libraries are in this folder `/usr/lib/x86_64-linux-gnu/`. This folder is among those in `gcc --print-search-dirs`, but it's not in `$LIBRARY_PATH`.

Executing `export LIBRARY_PATH=$LIBRARY_PATH:/usr/lib/x86_64-linux-gnu/` before installing pytorch solved the problem on my machine.

**Comment by soumith** [comment]
ok excellent repro. I'll be looking into how to fix the CMake to also look in that folder. I'll keep this issue open till i fix it.

**Comment by apaszke** [comment]
Fixed in #650.
